applications/main/nfc/plugins/supported_cards/microel.c

```diff
@@ -0,0 +1,195 @@
+                microel_1k_keys[i].b = nfc_util_bytes2num(keyB, KEY_LENGTH);
+            }
+        }
+
+        MfClassicDeviceKeys keys = {};
+        for(size_t i = 0; i < mf_classic_get_total_sectors_num(data->type); i++) {
+            nfc_util_num2bytes(microel_1k_keys[i].a, sizeof(MfClassicKey), keys.key_a[i].data);
+            FURI_BIT_SET(keys.key_a_mask, i);
+            nfc_util_num2bytes(microel_1k_keys[i].b, sizeof(MfClassicKey), keys.key_b[i].data);
+            FURI_BIT_SET(keys.key_b_mask, i);
+        }
+
+        error = mf_classic_poller_sync_read(nfc, &keys, data);
+        if(error != MfClassicErrorNone) {
+            FURI_LOG_W(TAG, "Failed to read data");
+            break;
+        }
+
+        nfc_device_set_data(device, NfcProtocolMfClassic, data);
+
+        is_read = mf_classic_is_card_read(data);
+    } while(false);
+
+    mf_classic_free(data);
+
+    return is_read;
+}
+
+static bool microel_parse(const NfcDevice* device, FuriString* parsed_data) {
+    furi_assert(device);
+    furi_assert(parsed_data);
+
+     const MfClassicData* data = nfc_device_get_data(device, NfcProtocolMfClassic);
+
+    bool parsed = false;
+
+    do {
+        //Get UID
+        size_t uid_len;
+        const uint8_t* uid = mf_classic_get_uid(data, &uid_len);
+        if(uid_len != UID_LENGTH) break;
+
+        // Generate key from uid
+        uint8_t keyA[KEY_LENGTH];
+        generateKeyA(uid, UID_LENGTH, keyA);
+
+        // Verify key
+        MfClassicSectorTrailer* sec_tr = mf_classic_get_sector_trailer_by_sector(data, verify_sector);
+        uint64_t key = nfc_util_bytes2num(sec_tr->key_a.data, 6);
+        uint64_t key_for_check_from_array = nfc_util_bytes2num(keyA, KEY_LENGTH);
+        if(key != key_for_check_from_array) return false;
+
+        //Get credit in block number 8
+        uint16_t balance = (data->block[4].data[6] << 8) | (data->block[4].data[5]);
+        uint16_t previus_balance = (data->block[5].data[6] << 8) | (data->block[5].data[5]);
+        furi_string_cat_printf(parsed_data, "\e#Microel Card\n");
+        furi_string_cat_printf(parsed_data, "UID:");
+        for(size_t i = 0; i < UID_LENGTH; i++) {
+            furi_string_cat_printf(parsed_data, " %02X", uid[i]);
+        }
+        furi_string_cat_printf(parsed_data, "\nCurrent Credit: %d.%02d E ", balance / 100, balance % 100);
+        furi_string_cat_printf(parsed_data, "\nPrevius Credit: %d.%02d E ", previus_balance / 100, previus_balance % 100);
+        furi_string_cat_printf(parsed_data, "\nTransaction number: %d \n", data->block[4].data[0]);
+        parsed = true;
+    } while(false);
+
+    return parsed;
+}
+
+/* Actual implementation of app<>plugin interface */
+static const NfcSupportedCardsPlugin microel_plugin = {
+    .protocol = NfcProtocolMfClassic,
+    .verify = NULL, // the verification I need is based on verifying the keys generated via uid and try to authenticate not like on mizip that there is default b0
```

Comment:
Verify function is necessary in this case because read function is quite long. It make reading experience worse because flipper tries to read any MFC card with microel keys, which takes a few seconds.

Can you generate one key based on UID for one sector, try to auth in verify() function? If auth succeed the read() function will be called, if not - then it's not microel card

Reply:
@gornekich is there a simple way to get the UID in this function? Because the only argument is the `Nfc` struct. We could execute [these](https://github.com/flipperdevices/flipperzero-firmware/blob/0789cbdefac0ac2f015059e896ee88c78aa9e492/lib/nfc/protocols/iso14443_3a/iso14443_3a_poller_i.c#L148) commands again, but I don't think we should manage this in plugins.

Reply:
Maybe we could put a simple check inside `read`?

Like this:
```c
// 1. Get data
MfClassicData* data = mf_classic_alloc();
nfc_device_copy_data(device, NfcProtocolMfClassic, data);

// 2. Check UID
size_t uid_len = 0;
const uint8_t* uid = mf_classic_get_uid(data, &uid_len);
FURI_LOG_D(TAG, "UID identified: %02X%02X%02X%02X", uid[0], uid[1], uid[2], uid[3]);
if(uid_len != UID_LENGTH) {
    return false;
}

// 3. Generate key
MfClassicKey key = {0};
generateKeyA(uid, UID_LENGTH, key.data);

// 4. Check key
const uint8_t block_num = mf_classic_get_first_block_num_of_sector(0);

MfClassicAuthContext auth_context;
MfClassicError error =
    mf_classic_poller_sync_auth(nfc, block_num, &key, MfClassicKeyTypeA, &auth_context);
if(error != MfClassicErrorNone) {
    return false;
}
// Other read stuff
```

Reply:
@giacomoferretti yes, your solution looks good!

Reply:
And sorry for late response. Please, update this PR with your solution